src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -1552,18 +1715,181 @@ CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
+}
+
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::EncryptPrivacy(MutableByteSpan & plaintext, const ByteSpan & nonce) const
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptPrivacy(MutableByteSpan & ciphertext, const ByteSpan & nonce) const
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+GroupDataProviderImpl::GroupSessionIterator * GroupDataProviderImpl::IterateGroupSessions(uint16_t session_id)
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mGroupSessionsIterator.CreateObject(*this, session_id);
+}
+
+GroupDataProviderImpl::GroupSessionIteratorImpl::GroupSessionIteratorImpl(GroupDataProviderImpl & provider, uint16_t session_id) :
+    mProvider(provider), mSessionId(session_id)
+{
+    FabricList fabric_list;
+    ReturnOnFailure(fabric_list.Load(provider.mStorage));
+    mFirstFabric = fabric_list.first_fabric;
+    mFabric      = fabric_list.first_fabric;
+    mFabricCount = 0;
+    mFabricTotal = fabric_list.fabric_count;
+    mMapCount    = 0;
+    mFirstMap    = true;
+}
+
+size_t GroupDataProviderImpl::GroupSessionIteratorImpl::Count()
+{
+    FabricData fabric(mFirstFabric);
+    size_t count = 0;
+
+    for (size_t i = 0; i < mFabricTotal; i++, fabric.fabric_index = fabric.next)
+    {
+        if (CHIP_NO_ERROR != fabric.Load(mProvider.mStorage))
+        {
+            break;
+        }
+
+        // Iterate key sets
+        KeyMapData mapping(fabric.fabric_index, fabric.first_map);
+
+        // Look for the target group in the fabric's keyset-group pairs
+        for (uint16_t j = 0; j < fabric.map_count; ++j, mapping.id = mapping.next)
+        {
+            if (CHIP_NO_ERROR != mapping.Load(mProvider.mStorage))
+            {
+                break;
+            }
+
+            // Group found, get the keyset
+            KeySet keyset;
+            if (CHIP_NO_ERROR != mProvider.GetKeySet(fabric.fabric_index, mapping.keyset_id, keyset))
+            {
+                break;
+            }
+            ByteSpan epoch_key = keyset.GetCurrentKey();
+            if (epoch_key.empty())
+            {
+                continue;
+            }
+
+            uint8_t operational_key[Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+            MutableByteSpan key_span(operational_key, sizeof(operational_key));
+
+            if (CHIP_NO_ERROR == Crypto::DeriveGroupOperationalKey(epoch_key, key_span))
+            {
+                uint16_t session_id = 0;
+                if (CHIP_NO_ERROR == Crypto::DeriveGroupSessionId(key_span, session_id) && session_id == mSessionId)
+                {
+                    count++;
+                }
+            }
+        }
+    }
+
+    return count;
+}
+
+bool GroupDataProviderImpl::GroupSessionIteratorImpl::Next(GroupSession & output)
+{
+    while (mFabricCount < mFabricTotal)
+    {
+        FabricData fabric(mFabric);
+        VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mProvider.mStorage), false);
```

Comment:
Since the key hash is 16-bit long, there is one chance in 65k of collision. I don't like those odds. However, we absolutely need to find ways to optimize the group data storage. I will consider this suggestion for future improvements, but I don't think I will implemented on this PR.